text-editor/text-editor-002-en.md

```diff
@@ -0,0 +1,227 @@
+The hard part was to produce the cleanest HTML markup as possible, eg for a
+`Hello world` text, when `Hello ` is already bold and we make `world` bold
+too, we initialy had
+```html
+<b>Hello </b><b>world</b>
+```
+two `b` tags following each other, but this should have been simplified to
+```html
+<b>Hello world</b>
+```
+This sanitization was done in a pass once the new `b` tag was added to the DOM.
+
+In the end, this was working fine but the number of lines of code started to
+grow, and we were discovering more and more edge cases which were painful to
+cover.
+
+Finally, we discovered a method which would have saved us all.
+
+### The holy grail : document.execCommand
+
+And then we discovered
+[`document.execCommand`](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand) \o/
+This function applies some text transformations to the current selection. Eg
+to bold / unbold the selection, just execute :
+```js
+document.execCommand('bold')
+```
+and that's it ! Your selection is wrapped inside a `b` tag, the produced
+markup is very clean (remember the previous step where we had to sanitize the
+markup manually).
+This is very neat, it enables rich text edition directly in the browser (and
+by the browser :p).
+Checkout [all the available commands](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand#Commands)
+to see what's possible to do !
+Needless to say that when we discovered this, we immediately stopped the
+previous attempt which was about manipulating the DOM by ourselves, and
+it was definitely time for a beer of victory !
+
+Considering the commands list, we had all the operators we needed to build
+our text editor. We only had to execute the right command when clicking on
+the right button.
+
+## Design
+
+The text editor is composed of two distinct sets of tools:
+- the paragraph toolbox (displayed when the cursor is on an empty paragraph)
+- and the text toolbox (displayed when a text is selected)
+
+![paragraph toolbox](/images/text-editor/paragraph-toolbox.png)
+*The paragraph toolbox*
+
+![text toolbox](/images/text-editor/text-toolbox.png)
+*The text toolbox*
+
+### Paragraph toolbox
+
+The paragraph toolbox enables some media insertion in a new paragraph. When
+clicking on the `image` button, it will open a media picker. We can then
+search the image to include from our database, or upload a new image.
+
+The `video` button behaves almost the same way than the `image` button :
+it will open the media picker for videos.
+
+For the tweet and youtube video insertion, a `text` input is displayed
+on the empty paragraph when such buttons are clicked to enter the URL of the
+tweet or the youtube video to insert. When submitting the form, the tweet
+or the video is inserted.
+
+### Text toolbox
+
+The text toolbox allows some text modifications :
+- turn the curent paragraph to a title
+- enable / disable italic
+- bold / unbold
+- enable / disable underline
+- insert a link
+- quote the selection
```

Comment:
I think this remeinder is not so helfull as you already introduced these elements on top of the article :)